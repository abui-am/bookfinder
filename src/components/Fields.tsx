/* eslint-disable import/prefer-default-export */
import clsx from 'clsx';
import { cloneElement, DetailedHTMLProps, InputHTMLAttributes } from 'react';

export const TextField: React.FC<
  DetailedHTMLProps<InputHTMLAttributes<HTMLInputElement>, HTMLInputElement> & {
    variant?: 'outlined' | 'contained';
    hasError?: boolean;
    Icon?: JSX.Element | ((props: { className: string }) => JSX.Element);
    EndIcon?: JSX.Element | ((props: { className: string }) => JSX.Element);
  }
> = ({ className, hasError, Icon, variant = 'outlined', EndIcon, ...props }) => {
  const variation = variant === 'outlined' ? 'border-gray-300 border' : 'bg-gray-100';
  const errorStyle = hasError ? 'ring-red-500 ring-inset border-transparent outline-none ring-2' : '';
  return (
    <div className="relative">
      {Icon && (
        <div className="absolute flex items-center left-3 top-0 bottom-0 m-auto text-blueGray-400">
          {typeof Icon === 'function'
            ? Icon({ className: 'text-gray-600' })
            : cloneElement(Icon, { className: 'text-gray-600' })}
        </div>
      )}

      <input
        {...props}
        className={clsx(
          Icon ? 'pl-11' : '',
          errorStyle,
          variation,
          'h-11 w-full rounded-full px-3 outline-none',
          'focus:ring-blue-600 focus:ring-inset focus:border-transparent focus:outline-none focus:ring-2',
          'transition-all duration-150 ease-in',
          className
        )}
      />

      {EndIcon && (
        <div className="absolute flex items-center top-0 bottom-0 right-3 m-auto text-blueGray-400">
          {typeof EndIcon === 'function'
            ? EndIcon({ className: 'text-gray-600' })
            : cloneElement(EndIcon, { className: 'text-gray-600' })}
        </div>
      )}
    </div>
  );
};
