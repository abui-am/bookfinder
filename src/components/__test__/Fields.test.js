import '@testing-library/jest-dom/extend-expect';

import { render, screen } from '@testing-library/react';
import user from '@testing-library/user-event';

import { TextField } from '../Fields';

describe('TextField', () => {
  it('should render properly', () => {
    render(<TextField />);
  });

  it('should call onChange when typing something', () => {
    const mockOnChange = jest.fn();
    render(<TextField aria-label="test" onChange={(e) => mockOnChange(e.target.value)} />);

    const el = screen.getByLabelText('test');
    user.type(el, 'random');
    expect(mockOnChange).toBeCalled();
    expect(mockOnChange).toBeCalledTimes(6);

    expect(mockOnChange).toHaveBeenNthCalledWith(1, 'r');
    expect(mockOnChange).toHaveBeenNthCalledWith(2, 'ra');
    expect(mockOnChange).toHaveBeenNthCalledWith(3, 'ran');
  });

  it('should call onKeyPress when typing something', () => {
    const mockOnChange = jest.fn();
    render(<TextField aria-label="test" onKeyDown={(e) => mockOnChange(e.key)} />);

    const el = screen.getByLabelText('test');
    user.type(el, '{enter}');
    expect(mockOnChange).toBeCalledWith('Enter');
  });

  it('should render Icon when given', () => {
    const el = <div>Testing</div>;
    render(<TextField Icon={el} />);
    const el2 = screen.getByText('Testing');
    expect(el2).toBeVisible();
    expect(el2).toHaveAttribute('class', 'text-gray-600');
  });

  it('should render EndIcon when given', () => {
    const endIcon = <div>Testingssssssssssssss</div>;
    render(<TextField EndIcon={endIcon} />);
    const el2 = screen.getByText('Testingssssssssssssss');
    expect(el2).toBeVisible();
    expect(el2).toMatchSnapshot();
  });

  it('should render snapshop', () => {
    render(
      <div aria-label="testing">
        <TextField />
      </div>
    );
    const test = screen.getByLabelText('testing');
    expect(test).toMatchSnapshot();
  });
});
