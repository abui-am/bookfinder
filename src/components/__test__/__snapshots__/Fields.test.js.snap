// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField should render EndIcon when given 1`] = `
<div
  class="text-gray-600"
>
  Testingssssssssssssss
</div>
`;

exports[`TextField should render snapshop 1`] = `
<div
  aria-label="testing"
>
  <div
    class="relative"
  >
    <input
      class="border-gray-300 border h-11 w-full rounded-full px-3 outline-none focus:ring-blue-600 focus:ring-inset focus:border-transparent focus:outline-none focus:ring-2 transition-all duration-150 ease-in"
    />
  </div>
</div>
`;
